[WGSL] Validate constraints on type variables when resolving overloads

tadeuzagallo opened the pull request:
#### 8f488e02e9be1fff9a916ce0d395b05e5bc0134b
<pre>
[WGSL] Validate constraints on type variables when resolving overloads
<a href="https://bugs.webkit.org/show_bug.cgi?id=254669">https://bugs.webkit.org/show_bug.cgi?id=254669</a>
rdar://problem/107370152

Reviewed by Myles C. Maxfield.

Type variables don&apos;t always accept any type, they usually have constraints such
as &quot;the type must be a floating point&quot; or the like. We were already specifying
the constraints, but not validating it. This patch modifies Overload::assign
to validate that a type actually satisfies the constraints before the assignment.

* Source/WebGPU/WGSL/Overload.cpp:
(WGSL::OverloadResolver::considerCandidate):
(WGSL::OverloadResolver::unify):
(WGSL::OverloadResolver::assign):
* Source/WebGPU/WGSL/Overload.h:
* Source/WebGPU/WGSL/tests/invalid/overload.wgsl:

Canonical link: <a href="https://commits.webkit.org/262328@main">https://commits.webkit.org/262328@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/297dbad6e3b330b1aef3e327caf04758d60160a8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1251 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1285 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1325 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2046 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1121 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1338 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1354 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1279 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1261 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1164 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1909 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1194 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1153 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1132 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1146 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1184 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2244 "268 api tests failed or timed out") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1194 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1094 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1131 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1163 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/314 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1221 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/Overload.cpp:
```diff
@@ -363,10 +362,83 @@ bool OverloadResolver::unify(const AbstractValue& parameter, unsigned argumentVa
 {
     logLn("assign ", variable, " => ", *type);
+    if (variable.constraints) {
+        auto* primitive = std::get_if<Types::Primitive>(type);
```
we generally use `downcast<>()` in WebKit instead; is there a benefit to `get_if`?

tadeuzagallo replied:
`get_if` is for variants, `downcast` is for subclass, and this type is a variant. @djg suggested moving to class and using `downcast` instead, but I haven't gotten around to it yet.

litherum commented on Source/WebGPU/WGSL/Overload.h:
```diff
@@ -38,20 +38,21 @@ using AbstractValue = std::variant<NumericVariable, unsigned>;
-        F32 = 1 << 1,
-        AbstractFloat = 1 << 2,
+
+        Bool          = 1 << 0,
```
Hey we have booleans now! Exciting!

litherum commented on Source/WebGPU/WGSL/Overload.cpp:
```diff
@@ -363,10 +362,83 @@ bool OverloadResolver::unify(const AbstractValue& parameter, unsigned argumentVa
+            if (!(variable.constraints & TypeVariable::F32))
+                return false;
+            break;
+        // FIXME: Add F16 support
+        // https://bugs.webkit.org/show_bug.cgi?id=254668
```
Is fp16 special somehow? I don't think it is. It's supported on all Metal devices as a first-class citizen. Seems like there's no reason not to just do it now 🤔

tadeuzagallo replied:
It's conditionally enabled by a directive, and we don't support directives yet. It should be trivial to implement, I'm just focusing on the default functionality for now.

litherum commented on Source/WebGPU/WGSL/Overload.cpp:
```diff
@@ -363,10 +362,83 @@ bool OverloadResolver::unify(const AbstractValue& parameter, unsigned argumentVa
+        if (!primitive)
+            return false;
+
+        switch (primitive->kind) {
+        case Types::Primitive::AbstractInt:
```
Maybe I'm not understanding it correctly, but this seems backwards to me. I'd expect the switch to switch over the constraint, and then for each constraint kind, ask if the primitive fits the constraint. Here, it looks like you're switching over the primitive type and then asking which constraints would fit the primitive. Is this intentional?

tadeuzagallo replied:
There a few things to consider:
- constraint is effectively a  bit set, so we can't simply switch on it
- a type might fit multiple constraints
- abstract types might have to be promoted to fit a constraint, and we have to choose the cheapest promotion which satisfies the constraint according to the conversion rank

Given all that, I thought it was easier done this way, but it might be possible to invert it.

tadeuzagallo committed:
[WGSL] Validate constraints on type variables when resolving overloads
https://bugs.webkit.org/show_bug.cgi?id=254669
rdar://problem/107370152

Reviewed by Myles C. Maxfield.

Type variables don't always accept any type, they usually have constraints such
as "the type must be a floating point" or the like. We were already specifying
the constraints, but not validating it. This patch modifies Overload::assign
to validate that a type actually satisfies the constraints before the assignment.

* Source/WebGPU/WGSL/Overload.cpp:
(WGSL::OverloadResolver::considerCandidate):
(WGSL::OverloadResolver::unify):
(WGSL::OverloadResolver::assign):
* Source/WebGPU/WGSL/Overload.h:
* Source/WebGPU/WGSL/tests/invalid/overload.wgsl:

Canonical link: https://commits.webkit.org/262328@main
